src/mrdsrg-spin-adapted/sa_mrpt2.cc

```diff
@@ -124,24 +125,63 @@ void SA_MRPT2::build_minimal_V() {
     auto dim2 = na * nc;
     auto dim1 = nv * dim2;
 
-    auto Bsub = ambit::Tensor::build(tensor_type_, "Bsub PT2", {nQ, nv});
-    auto Vsub = ambit::Tensor::build(tensor_type_, "Vsub PT2", {na, nv, na});
+    auto Baa = B.block("Laa");
+
+    // to minimize the number of calls of ints_->three_integral_block,
+    // we store as many (Q|em) as possible in memory.
+    size_t max_num = dsrg_mem_.available() * 0.8 / (sizeof(double) * (nQ * nv + na * na * nv));
+
+    // separate core indices into batches
+    auto core_batches = split_indices_to_batches(core_mos_, max_num);
+    size_t nbatch = core_batches.size();
 
-    for (size_t c = 0; c < nc; ++c) {
-        Bsub.data() = ints_->three_integral_block(aux_mos_, virt_mos_, {core_mos_[c]}).data();
+    for (size_t batch = 0, offset = 0; batch < nbatch; ++batch) {
+        auto size = core_batches[batch].size();
+        auto Bsub = ints_->three_integral_block(aux_mos_, virt_mos_, core_batches[batch]);
 
-        Vsub("uev") = B.block("Laa")("guv") * Bsub("ge");
+        auto Vsub = ambit::Tensor::build(tensor_type_, "Vsub PT2", {na, na, nv, size});
+        Vsub("uvem") = Baa("guv") * Bsub("gem");
 
         Vsub.citerate([&](const std::vector<size_t>& i, const double& value) {
-            Vavac[i[0] * dim1 + i[1] * dim2 + i[2] * nc + c] = value;
+            Vavac[i[0] * dim1 + i[2] * dim2 + i[1] * nc + i[3] + offset] = value;
```

Comment:
I noticed a indexing change here. Is this due to a reorganization of the data structure?

Reply:
You are right. Previously, `Vsub` is arranged as `{na, nv, na}` but now is `{na, na, nv, size}`: a switch between second (`i[1]`) and third (`i[2]`) indices.